File: LeetCodeRecord/MethodFirst/560-SubarraySum.java
package MethodFirst;

import java.util.HashMap;
import java.util.Map;

class SubarraySum {
    public int subarraySum(int[] nums, int k) {
        int count = 0;

        for (int i = 0; i < nums.length; i++) {
            int sum = 0;

            for (int j = i + 1; j <= nums.length; j++) {
                sum += nums[j - 1];

                if (sum == k) {
                    count++;
                }
            }
        }

        return count;
    }    

    public int subarraySum2(int[] nums, int k) {
        int len = nums.length;
        int[] preSums = new int[len + 1];

        preSums[0] = 0;
        for (int i = 1; i <= len; i++) {
            preSums[i] = preSums[i - 1] + nums[i - 1];
        }

        int count = 0;
        for (int left = 0; left < len; left++) {
            for (int right = left + 1; right <= len; right++) {
                if (preSums[right] - preSums[left] == k) {
                    count++;
                }
            }
        }
        
        return count;
    }
    
    public int subarraySum3(int[] nums, int k) {
        int preSum = 0;

        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        map.put(0, 1);

        int count = 0;
        for (int i = 0; i < nums.length; i++) {
            preSum += nums[i];

            if (map.containsKey(preSum - k)) {
                count += map.get(preSum - k);
            } 

            if (map.containsKey(preSum)) {
                map.put(preSum, map.get(preSum) + 1);
            } else {
                map.put(preSum, 1);
            }
        }

        return count;
    }
}

File: LeetCodeRecord/ACFirst-Golang/560-SubarraySum.go
package main

func subarraySum(nums []int, k int) int {
	numsLen := len(nums)
	count := 0

	for i := 0; i < numsLen; i++ {
		sum := 0
		for j := i; j < numsLen; j++ {
			sum += nums[j]

			if sum == k {
				count++
			}
		}
	}

	return count
}

func subarraySum2(nums []int, k int) int {
	numsLen := len(nums)
	preSum := make([]int, numsLen+1)
	count := 0

	preSum[0] = 0
	for i := 0; i < numsLen; i++ {
		preSum[i+1] = preSum[i] + nums[i]
	}

	for i := 0; i < numsLen; i++ {
		for j := i + 1; j <= numsLen; j++ {
			if preSum[j]-preSum[i] == k {
				count++
			}
		}
	}

	return count
}

func subarraySum3(nums []int, k int) int {
	hashMap := make(map[int]int)
	hashMap[0] = 1
	preSum := 0
	count := 0

	for i := 0; i < len(nums); i++ {
		preSum += nums[i]

		if elem, ok := hashMap[preSum-k]; ok == true {
			count += elem
		}

		if elem, ok := hashMap[preSum]; ok == true {
			hashMap[preSum] = elem + 1
		} else {
			hashMap[preSum] = 1
		}
	}

	return count
}
